src/style/parse.js

```diff
@@ -35,6 +35,11 @@ styfn.parse = function( name, value, propIsBypass, propIsFlat ){
     }
   }
 
+  // cache the original, unparsed value for comparison on the next iteration
+  if( ret ){
+    ret.previousValue = util.copy( value );
```

Comment:
Is the copying strictly necessary?  It adds expense.

It's certainly not necessary in the non-mapper cases.

Reply:
I was making an assumption that the un-parsed value could be a type (ex: array) that would require copying. But if that's not the case, I can certainly remove it!

Alternatively and/or in addition to, another option would be to move this statement to [```applyParsedProperty```:](https://github.com/cytoscape/cytoscape.js/blob/6feca07b2c6f422a41993bcdc16fe3fc4dae3995/src/style/apply.js#L588)
```
case types.fn:
  ...
  flatProp = this.parse( prop.name, fnRetVal, prop.bypass, flatPropMapping );
  flatProp.previousValue = fnRetVal; // or util.copy(fnRetVal)
```

Then regardless of whether copying is required or not, only previous values for mapper cases get cached. Which would make sense as that's the only case where it gets used.

Reply:
If the goal is to avoid the expense of `applyParsedProperty()` for function mappers, could we use an early exit condition within `applyParsedProperty()`?  Then all the logic for this purpose would be contained in the same function.

Additionally, the top-level `parse()` function is just about caching the properties to avoid parsing expense.  It's the `parseImpl()` function that contains all of the real parsing logic.

It's also worth considering that line 39 would copy for all property types, whereas the code in your comment is better scoped:  Copying happens only for function property values in your comment.